lxd/seccomp/seccomp.go

```diff
@@ -1875,6 +1885,197 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
+		return int(-C.EPERM)
+	}
+
+	// has CAP_SYS_MODULE capability?
+	if !C.lxd_pid_cap_is_set(C.int(targetPID), C.CAP_SYS_MODULE, C.CAP_EFFECTIVE) {
+		ctx["err"] = "requestor process creds lacks CAP_SYS_MODULE"
+		return int(-C.EPERM)
+	}
+
+	moduleFileFD, err := unix.Openat(siov.procFd, fmt.Sprintf("fd/%d", fd), unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file (unix.Openat): %s", err)
+		return int(-C.EPERM)
+	}
+
+	moduleFile := os.NewFile(uintptr(moduleFileFD), "/proc/<pid>/fd/<fd>")
+	if moduleFile == nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file (os.NewFile): %s", err)
+		return int(-C.EPERM)
+	}
+
+	defer func() { _ = moduleFile.Close() }()
+
+	forksyscallgoargs := []string{
+		"forksyscallgo",
+		"finit_module_parse",         // <syscall_operation>
+		fmt.Sprintf("%d", targetPID), // <PID>
+		fmt.Sprintf("%d", 0),         // <PidFd>
+		fmt.Sprintf("%d", 3),         // <module_fd>
+	}
+
+	var buffer bytes.Buffer
+	var output bytes.Buffer
+	p := subprocess.NewProcessWithFds(util.GetExecPath(), forksyscallgoargs, nil,
+		&nullWriteCloser{&output}, &nullWriteCloser{&buffer})
+
+	// We want to drop privileges, as we use "debug/elf" Golang
+	// package which is considered as not safe for production use.
+	// By spawning an unprivileged process we can safely use this package.
+	p.SetCreds(s.s.OS.UnprivUID, s.s.OS.UnprivGID)
+
+	// Let's set timeout for the helper process.
+	// It's just a precautionary measure to prevent blocakge of a syscall interception
+	// daemon if anything goes wrong inside the debug/elf Golang package internals.
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	err = p.StartWithFiles(timeoutCtx, []*os.File{moduleFile})
+	if err != nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file: %s", err)
+		return int(-C.EPERM)
+	}
+
+	exitCode, err := p.Wait(context.TODO())
+	if err != nil || exitCode != 0 {
+		ctx["err"] = fmt.Sprintf("forksyscallgo finit_module_parse helper failed. Err: %s exitCode: %d stdout: %s stderr: %s",
+			err, exitCode, output.String(), buffer.String())
+		return int(-C.EPERM)
+	}
+
+	moduleName := output.String()
+	ctx["module_name"] = moduleName
+
+	// Check that this module is in the allowlist
+	inAllowList := false
+	kernelModules := c.ExpandedConfig()["linux.kernel_modules"]
+	if kernelModules != "" {
+		for _, module := range strings.Split(kernelModules, ",") {
+			module = strings.TrimPrefix(module, " ")
+
+			if module == moduleName {
+				inAllowList = true
+			}
+		}
+	}
+
+	if !inAllowList {
+		ctx["err"] = fmt.Sprintf("Module %s is not in allowlist (linux.kernel_modules)", moduleName)
+		return int(-C.EPERM)
+	}
+
+	if shared.PathExists(fmt.Sprintf("/sys/module/%s", moduleName)) {
+		return int(-C.EEXIST)
+	}
+
+	err = util.LoadModule(moduleName)
+	if err != nil {
+		ctx["err"] = fmt.Sprintf("Failed to load module %s", moduleName)
```

Comment:
Please can you use %q when including strings in errors (so it quotes the contents), also please can you include the `err` value in the returned error, so something like this: `fmt.Sprintf("Failed to load module %q: %v", moduleName, err)`

Reply:
Fixed!